demos/canvas/app/src/hooks/useActivityToast.tsx

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Toaster } from "@blueprintjs/core";
+import type { ActivityEvent } from "@demo/canvas.sdk";
+import { ActivityEventModel } from "@demo/canvas.sdk";
+import type { DocumentRef } from "@palantir/pack.document-schema.model-types";
+import { type RefObject, useCallback, useEffect, useRef, useState } from "react";
+import { ShapeUpdateToast } from "../components/toast/ShapeUpdateToast.js";
+
+export function useActivityToast(
+  docRef: DocumentRef,
+  toasterRef: RefObject<Toaster | null>,
+): void {
+  const updateCountsRef = useRef<Map<string, { count: number; key: string }>>(new Map());
+  const [, forceUpdate] = useState({});
+
+  const triggerUpdate = useCallback(() => {
+    forceUpdate({});
+  }, []);
```

Comment:
what's this for?

Reply:
oh thanks! it was originally for some bad effect/ref usage further up the tree but tidied that up now too